doc/proposals/composable-integration-in sdk.md

```diff
@@ -0,0 +1,173 @@
+
+### User Story 1
+
+As an operator developer, I wish not to have the value of a field be hard-wired. These are typically
+values that are computed dynamically when an application is deployed. For example, the admin URL for a Kafka
+deployment is known only after Kafka has been successfully deployed. If my CRD requires the URL, then
+rather that requiring a string value for that field, I can define it to be a reference to another object
+(perhaps the data is in the status field of the Kubernetes object managing Kafka).
+
+### User Story 2
+
+Often the yamls for a collection of resources require the same data in many places (e.g., a host name or port number).
+Rather than hardwiring the same value everywhere, I can define cross-resource references to a unique configmap that contains 
+the data. This makes it easier to change the data if needed, and can allow all resources to be configured
+dynamically (i.e. changes in the configmap are absorbed and resource that point to it are updated).
+
+
+### Implementation Details/Notes/Constraints (optional)
+
+The Composable SDK offers the following types.
+
+```golang
+type ObjectRef struct {
+	GetValueFrom ComposableGetValueFrom `json:"getValueFrom"`
+}
+
+type ComposableGetValueFrom struct {
+	Kind               string   `json:"kind"`
+	APIVersion         string   `json:"apiVersion,omitempty"`
+	Name               string   `json:"name,omitempty"`
+	Labels             []string `json:"labels,omitempty"`
+	Namespace          string   `json:"namespace,omitempty"`
+	Path               string   `json:"path"`
+	FormatTransformers []string `json:"format-transformers,omitempty"`
+}
+```
+
+An `ObjectRef` can be used to specify the type of any field of a CRD definition, allowing the value to be determined dynamically.
+For a detailed explanation of how to specify an object reference according to this schema, see [here](https://github.com/IBM/composable/blob/master/README.md#getvaluefrom-elements).
+
+The Composable SDK offers the following function for resolving the value of cross-resource references.
+
+```golang
+func ResolveObject(r client.Client, config *rest.Config, object interface{}, resolved interface{}, composableNamespace string) *ComposableError 
+```
+
+The function `ResolveObject` takes a Kubernetes client and configuration, and `object` to resolve, and a blank object
+`resolved` that will contain the result, as well as a namespace. The namespace is the default used when references
+do not specify one. This function will cast the result to the type of the `resolved` object, provided that
+appropriate data transformers have been included in the reference definitions (see [tutorial](https://github.com/IBM/composable/blob/master/sdk/docs/tutorial.md) for an example).
+
+The `ResolveObject` function uses caching for looking up object kinds, as well as for the objects themselves, in order
+to ensure that a consistent view of the data is obtained. If any data is not available at the time of the lookup,
```

Comment:
ResolveObject uses a cache of its own to read objects only once. If there are 2 references to different parts of the same object, it does not read the object twice. It will cache the first read and use the result to fulfill the second reference. 

Consistency means that if the reference object is updated in the mean time, then ResolveObject sees a "consistent view" of it, because it only reads it once. It doesn't see intermediate updates (that could happen if it read the object multiple times to get different parts of it).